src/pandas_openscm/index_manipulation.py

```diff
@@ -714,3 +717,188 @@ def update_levels_from_other(
+
+    Returns
+    -------
+    :
+        New MultiIndex with the levels set to the provided values
+
+    Raises
+    ------
+        TypeError
+            If `ini` is not a MultiIndex
+        ValueError
+            If the length of the values is not equal to the length of the index
+
+    Examples
+    --------
+    >>> start = pd.MultiIndex.from_tuples(
+    ...     [
+    ...         ("sa", "ma", "v1", "kg"),
+    ...         ("sb", "ma", "v2", "m"),
+    ...         ("sa", "mb", "v1", "kg"),
+    ...         ("sa", "mb", "v2", "m"),
+    ...     ],
+    ...     names=["scenario", "model", "variable", "unit"],
+    ... )
+    >>> start
+    MultiIndex([('sa', 'ma', 'v1', 'kg'),
+                ('sb', 'ma', 'v2',  'm'),
+                ('sa', 'mb', 'v1', 'kg'),
+                ('sa', 'mb', 'v2',  'm')],
+               names=['scenario', 'model', 'variable', 'unit'])
+    >>>
+    >>> # Set a new level with a single string
+    >>> set_levels(
+    ...     start,
+    ...     {"new_variable": "xyz"},
+    ... )
+    MultiIndex([('sa', 'ma', 'v1', 'kg', 'xyz'),
+            ('sb', 'ma', 'v2',  'm', 'xyz'),
+            ('sa', 'mb', 'v1', 'kg', 'xyz'),
+            ('sa', 'mb', 'v2',  'm', 'xyz')],
+           names=['scenario', 'model', 'variable', 'unit', 'new_variable'])
+    >>>
+    >>> # Replace a level with a collection
+    >>> set_levels(
+    ...     start,
+    ...     {"new_variable": [1, 2, 3, 4]},
+    ... )
+    MultiIndex([('sa', 'ma', 'v1', 'kg', 1),
+                ('sb', 'ma', 'v2',  'm', 2),
+                ('sa', 'mb', 'v1', 'kg', 3),
+                ('sa', 'mb', 'v2',  'm', 4)],
+               names=['scenario', 'model', 'variable', 'unit', 'new_variable'])
+    >>>
+    >>> # Replace a level with a single value and add a new level
+    >>> set_levels(
+    ...     start,
+    ...     {"model": "new_model", "new_variable": ["xyz", "xyz", "x", "y"]},
+    ... )
+    MultiIndex([('sa', 'new_model', 'v1', 'kg', 'xyz'),
+                ('sb', 'new_model', 'v2',  'm', 'xyz'),
+                ('sa', 'new_model', 'v1', 'kg',   'x'),
+                ('sa', 'new_model', 'v2',  'm',   'y')],
+               names=['scenario', 'model', 'variable', 'unit', 'new_variable'])
+    """
+    levels: list[pd.Index[Any]] = list(ini.levels)
+    codes: list[list[int] | npt.NDArray[np.integer[Any]]] = list(ini.codes)
+    names: list[str] = list(ini.names)
+
+    for level, value in levels_to_set.items():
+        if isinstance(value, Collection) and not isinstance(value, str):
+            if len(value) != len(ini):
+                msg = (
+                    f"Length of values for level '{level}' does not "
+                    f"match index length: {len(value)} != {len(ini)}"
+                )
+                raise ValueError(msg)
+            new_level, new_codes = create_level_from_collection(level, value)
+        else:
+            new_level = pd.Index([value], name=level)
+            new_codes = [0] * ini.shape[0]
```

Comment:
```suggestion
            new_codes = np.zeros(ini.shape[0])
```

Micro-optimisation, but using numpy directly where it's easy is a good habit to build